### PR TITLE
Dreamview: fix synchronized point cloud rendering in dreamview

### PR DESCRIPTION
**Pull request by weidezhang**
1. synchronize timestamp when simulation world pushes to frontend
2. eliminate front end point cloud timer trigger
3. more smoother visualization

### CODE REVIEW

File: modules/dreamview/backend/handlers/websocket_handler.h [hunk 1]
```diff
@@ -66,6 +66,7 @@ class WebSocketHandler : public CivetWebSocketHandler {
    * @returns true to keep socket open, false to close it
    */
   bool handleConnection(CivetServer *server, const Connection *conn) override {
+    AINFO<<"web socket establishes a connection";
```
Comment (vlin17) [inline modules/dreamview/backend/handlers/websocket_handler.h]: lint: spaces  "AINFO << "

File: modules/dreamview/frontend/src/store/websocket/websocket_point_cloud.js [hunk 2]
```diff
@@ -37,22 +38,27 @@ export default class PointCloudWebSocketEndpoint {
                 RENDERER.updatePointCloud({num:[]});
               }
             } else if (STORE.options.showPointCloud === true && event.data.num !== undefined) {
+                console.log('render point cloud data');
```
Comment (vlin17) [inline modules/dreamview/frontend/src/store/websocket/websocket_point_cloud.js:41]: do we still need this log?

File: modules/dreamview/frontend/src/store/websocket/websocket_realtime.js [hunk 3]
```diff
@@ -16,7 +16,8 @@ export default class RealtimeWebSocketEndpoint {
         this.routingTime = undefined;
         this.currentMode = null;
         this.worker = new Worker();
-
+        this.cameraWS = null;
```
Comment (vlin17) [inline modules/dreamview/frontend/src/store/websocket/websocket_realtime.js:19]: Seems like the change for this.cameraWS is half way. Can we either finish it or remove the reference here?

File: modules/dreamview/backend/point_cloud/point_cloud_updater.h [hunk 4]
```diff
@@ -52,8 +53,11 @@ class PointCloudUpdater {
    * @brief Constructor with the websocket handler.
    * @param websocket Pointer of the websocket handler that has been attached to
    * the server.
+   * @param simulationworldupater pointer
```
Comment (vlin17) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.h:56]: upater => updater

File: modules/dreamview/backend/handlers/websocket_handler.h [hunk 5]
```diff
@@ -29,7 +29,7 @@
 
 #include "CivetServer.h"
 #include "third_party/json/json.hpp"
-
+#include "cyber/common/log.h"
```
Comment (vlin17) [inline modules/dreamview/backend/handlers/websocket_handler.h:32]: is this needed still?

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 6]
```diff
@@ -168,23 +151,57 @@ void PointCloudUpdater::UpdatePointCloud(
...
-      new pcl::PointCloud<pcl::PointXYZ>);
-  voxel_grid.filter(*pcl_filtered_ptr);
-  AINFO << "filtered point cloud data size: " << pcl_filtered_ptr->size();
+        new pcl::PointCloud<pcl::PointXYZ>);
+
+  /** by default, disable voxel filter since it's taking more than 500ms 
+      ideally the most efficient sampling method is to 
+      use per beam random sample for organized cloud(TODO)
+  **/
+  if (apply_voxel_filter) {
```
Comment (vlin17) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:195]: Do we have any concern about using "enable_voxel_filter_" here ?

File: modules/dreamview/frontend/src/store/websocket/websocket_realtime.js (hunk 3 above)
Comment (weidezhang) [reply modules/dreamview/frontend/src/store/websocket/websocket_realtime.js:19]: will remove it for now. will fix later.

File: modules/dreamview/backend/dreamview.cc [hunk 7]
```diff
@@ -57,8 +57,11 @@ Status Dreamview::Init() {
   // javascripts and handles websocket requests.
   std::vector<std::string> options = {
       "document_root",      FLAGS_static_file_dir,   "listening_ports",
-      FLAGS_server_ports,   "websocket_timeout_ms",  FLAGS_websocket_timeout_ms,
-      "request_timeout_ms", FLAGS_request_timeout_ms};
+      FLAGS_server_ports,   "websocket_timeout_ms",  "3600",
+      "request_timeout_ms", "3600",
```
Comment (unacao) [inline modules/dreamview/backend/dreamview.cc:61]: why change the gflag to literal for the two timeout settings above? 3600ms is also a weird choice

File: modules/dreamview/backend/handlers/BUILD [hunk 8]
```diff
@@ -8,6 +8,7 @@ cc_library(
     hdrs = ["websocket_handler.h"],
     copts = ['-DMODULE_NAME=\\"dreamview\\"'],
     deps = [
+        "//cyber",
```
Comment (unacao) [inline modules/dreamview/backend/handlers/BUILD:11]: why is this needed? any more specific dependency other than //cyber?

File: modules/dreamview/backend/handlers/websocket_handler.cc [hunk 9]
```diff
@@ -129,7 +129,7 @@ bool WebSocketHandler::SendData(Connection *conn, const std::string &data,
   int ret = 0;
   PERF_BLOCK(absl::StrCat(name_, ": Writing ", data.size(),
                           " bytes via websocket took"),
-             0.1) {
+             0.02) {
```
Comment (unacao) [inline modules/dreamview/backend/handlers/websocket_handler.cc:132]: reducing the threshold to 0.02s will cause too frequent logs in production setting

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 10]
```diff
@@ -84,12 +86,6 @@ void PointCloudUpdater::RegisterMessageHandlers() {
       "RequestPointCloud",
       [this](const Json &json, WebSocketHandler::Connection *conn) {
         std::string to_send;
-        // If there is no point_cloud data for more than 2 seconds, reset.
```
Comment (unacao) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:87]: why is this removed?

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 11]
```diff
@@ -168,23 +151,57 @@ void PointCloudUpdater::UpdatePointCloud(
...
-                         static_cast<float>(FLAGS_voxel_filter_size),
-                         static_cast<float>(FLAGS_voxel_filter_height));
+
   pcl::PointCloud<pcl::PointXYZ>::Ptr pcl_filtered_ptr(
-      new pcl::PointCloud<pcl::PointXYZ>);
-  voxel_grid.filter(*pcl_filtered_ptr);
-  AINFO << "filtered point cloud data size: " << pcl_filtered_ptr->size();
+        new pcl::PointCloud<pcl::PointXYZ>);
+
+  /** by default, disable voxel filter since it's taking more than 500ms
```
Comment (unacao) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:191]: please use comment format:
/*
my comments
*/

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 12]
```diff
@@ -168,23 +151,57 @@ void PointCloudUpdater::UpdatePointCloud(
...
+  }
+  pcl::PointCloud<pcl::PointXYZ>::Ptr pcl_ptr;
+  // Check if last filter process has finished before processing new data.
+  if (enable_voxel_filter_) {
+    if (future_ready_) {
+       future_ready_ = false;
+       // transform from drivers::PointCloud to pcl::PointCloud
+       pcl_ptr = ConvertPCLPointCloud(point_cloud);
+       std::future<void> f =
+      cyber::Async(&PointCloudUpdater::FilterPointCloud, this, pcl_ptr);
```
Comment (unacao) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:176]: fix indentation

File: modules/dreamview/backend/simulation_world/simulation_world_updater.h [hunk 13]
```diff
@@ -138,6 +142,10 @@ class SimulationWorldUpdater {
   boost::shared_mutex mutex_;
 
   std::unique_ptr<cyber::Timer> timer_;
+
+  volatile double last_pushed_adc_timestamp_sec_ = 0.0f;
+
+  WebSocketHandler::Connection *conn_;
```
Comment (unacao) [inline modules/dreamview/backend/simulation_world/simulation_world_updater.h:148]: what is this new variable used for?

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 14]
```diff
@@ -168,23 +151,57 @@ void PointCloudUpdater::UpdatePointCloud(
...
+}
+
+void PointCloudUpdater::UpdatePointCloud(
+    const std::shared_ptr<drivers::PointCloud> &point_cloud) {
+  if (!enabled_) {
+    return;
+  }
+  last_point_cloud_time_ = point_cloud->header().timestamp_sec();
+  if (simworld_updater_->LastAdcTimestampSec() == 0.0 ||
+    simworld_updater_->LastAdcTimestampSec()-last_point_cloud_time_ > 0.1) {
```
Comment (unacao) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:164]: "()-last_point_cloud_time_" reformat "() - last_point_cloud_time_"

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc (hunk 10 above)
Comment (weidezhang) [reply modules/dreamview/backend/point_cloud/point_cloud_updater.cc:87]: oh, i guess  we dont need to send empty point cloud if there is no update. i will use a flag to check if the current point cloud has been sent before and if so, skip send anything.  what do u think ?

Comment (weidezhang) [reply modules/dreamview/backend/point_cloud/point_cloud_updater.cc:87]: also i remove the localization estimate here since the timestamp is synced from simulation world

Comment (unacao) [reply modules/dreamview/backend/point_cloud/point_cloud_updater.cc:87]: this is for cleaning up point cloud if the user plays different rosbags with/without point clouds

**Comment by unacao** [comment]
please commit fixes as new commits instead of amend so that we can see the difference. all the commits in this PR can be squashed later when merging into master

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 15]
```diff
@@ -168,23 +156,58 @@ void PointCloudUpdater::UpdatePointCloud(
...
+  }
+  pcl::PointCloud<pcl::PointXYZ>::Ptr pcl_ptr;
+  // Check if last filter process has finished before processing new data.
+  if (enable_voxel_filter_) {
+    if (future_ready_) {
+       future_ready_ = false;
+       // transform from drivers::PointCloud to pcl::PointCloud
+       pcl_ptr = ConvertPCLPointCloud(point_cloud);
+       std::future<void> f =
+       cyber::Async(&PointCloudUpdater::FilterPointCloud, this, pcl_ptr);
```
Comment (unacao) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:181]: need 4 more spaces here

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc [hunk 16]
```diff
@@ -168,23 +156,58 @@ void PointCloudUpdater::UpdatePointCloud(
...
+void PointCloudUpdater::UpdatePointCloud(
+    const std::shared_ptr<drivers::PointCloud> &point_cloud) {
+  if (!enabled_) {
+    return;
+  }
+  last_point_cloud_time_ = point_cloud->header().timestamp_sec();
+  if (simworld_updater_->LastAdcTimestampSec() == 0.0 ||
+    simworld_updater_->LastAdcTimestampSec() - last_point_cloud_time_ > 0.1) {
+      AWARN << "skipping outdated point cloud data";
+      return;
```
Comment (unacao) [inline modules/dreamview/backend/point_cloud/point_cloud_updater.cc:171]: reduce 2 spaces for the two lines above

File: modules/dreamview/backend/point_cloud/point_cloud_updater.cc (hunk 10 above)
Comment (weidezhang) [reply modules/dreamview/backend/point_cloud/point_cloud_updater.cc:87]: sure i will restore the old logic.
